File: java/src/main/java/shapes/client/shell/InputReader.java
package shapes.client.shell;

import java.util.Scanner;

final class InputReader {
    private static final Scanner scanner = new Scanner(System.in);

    static String readInput() {
        System.out.print("Your Answer >>> ");
        String userInput = scanner.nextLine();
        userInput = userInput.trim().toLowerCase();
        readExit(userInput);
        return userInput;
    }

    static double readNumber() {
        while (true) {
            String userInput = readInput();

            try {
                return Double.parseDouble(userInput);
            } catch (NumberFormatException e) {
                System.out.println("Invalid number. Please try again.");
            }
        }
    }

    private static void readExit(String userInput) {
        if (userInput.equals("exit") || userInput.equals("e")) {
            System.out.println("Thank you for using the Shapes CLI!");
            System.exit(0);
        }
    }
}

File: haskell/src/Shapes/Client/Shell/InputReader.hs
module Shapes.Client.Shell.InputReader (readInput, readNumber) where

import Data.Char (toLower)
import System.Exit (exitSuccess)
import System.IO (hFlush, stdout)

readInput :: IO String
readInput = do
    putStr "Your Answer >>> "
    hFlush stdout
    userInput <- fmap (map toLower . trim) getLine
    readExit userInput
    return userInput

readNumber :: IO Double
readNumber = do
    userInput <- readInput
    case reads userInput :: [(Double, String)] of
        [(number, "")] -> return number
        _ -> do
            putStrLn "Invalid number. Please try again."
            readNumber

readExit :: String -> IO ()
readExit input
    | input `elem` ["exit", "e"] = do
        putStrLn "Thank you for using the Shapes CLI!"
        exitSuccess
    | otherwise = return ()

trim :: String -> String
trim = unwords . words
